jit/src/lib.rs

```diff
@@ -143,3 +211,54 @@ impl fmt::Debug for CompiledCode {
         f.write_str("[compiled code]")
     }
 }
+
+pub struct ArgsBuilder<'a> {
+    values: Vec<MaybeUninit<UnTypedAbiValue>>,
+    initialised: Vec<bool>,
+    code: &'a CompiledCode,
+}
+
+impl<'a> ArgsBuilder<'a> {
+    fn new(code: &'a CompiledCode) -> ArgsBuilder<'a> {
+        ArgsBuilder {
+            values: vec![MaybeUninit::uninit(); code.sig.args.len()],
+            initialised: vec![false; code.sig.args.len()],
+            code,
+        }
+    }
+
+    pub fn set(&mut self, idx: usize, value: AbiValue) {
+        self.values[idx] = MaybeUninit::new(value.to_untyped());
+        self.initialised[idx] = true;
+    }
+
+    pub fn is_set(&self, idx: usize) -> bool {
+        self.initialised[idx]
+    }
+
+    pub fn into_args(self) -> Option<Args<'a>> {
+        if self.initialised.iter().all(|x| *x) {
+            // SAFETY: we know all elements are initialised so it's safe to transmute away the
+            // MaybeUninit wrapper.
+            let values = unsafe {
+                std::mem::transmute::<Vec<MaybeUninit<UnTypedAbiValue>>, Vec<UnTypedAbiValue>>(
+                    self.values,
+                )
+            };
+            let cif_args = values.iter().map(|v| libffi::middle::Arg::new(v)).collect();
```

Comment:
If you're going to recollect the vec anyway, you may as well just `v.assume_init()` in the mapping function and avoid the transmute.

Reply:
I can't use `assume_init` as I need the reference into the `values` vector, `cif_args` is effectively a vector of pointers that are passed to `libffi`. This would be nicer if we had either https://github.com/rust-lang/rust/issues/63291 or https://github.com/rust-lang/rust/issues/63568.

Reply:
Ah, gotcha